plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -0,0 +1,493 @@
+  ## telemetry will be exposed by Intel PowerStat plugin.
+  ## Supported options:
+  ## "cpu_frequency", "cpu_c1_state_residency", "cpu_c6_state_residency", "cpu_busy_cycles", "cpu_temperature", "cpu_busy_frequency"
+  # cpu_metrics = []
+`
+}
+
+// Description returns a one-sentence description on the plugin.
+func (p *PowerStat) Description() string {
+	return `Intel PowerStat plugin enables monitoring of platform metrics (power, TDP) and Core metrics like temperature, power and utilization.`
+}
+
+// Gather takes in an accumulator and adds the metrics that the Input gathers.
+func (p *PowerStat) Gather(acc telegraf.Accumulator) error {
+	p.addGlobalMetrics(acc)
+
+	if p.perCoreMetrics && len(p.msr.GetCpuIDs()) > 0 {
+		p.addPerCoreMetrics(acc)
+	}
+
+	// Gathering the first iteration of metrics was skipped for most of them because they are based on delta calculations.
+	p.skipFirstIteration = false
+
+	return nil
+}
+
+func (p *PowerStat) addPerCoreMetrics(acc telegraf.Accumulator) {
+	var wg sync.WaitGroup
+	cpuIDs := p.msr.GetCpuIDs()
+	wg.Add(len(cpuIDs))
+
+	for _, cpuID := range cpuIDs {
+		go p.addMetricsForSingleCore(cpuID, acc, &wg)
+	}
+
+	wg.Wait()
+}
+
+func (p *PowerStat) addMetricsForSingleCore(cpuID string, acc telegraf.Accumulator, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	if p.coreCPUFrequency {
+		p.addCPUFrequencyMetric(cpuID, acc)
+	}
+
+	// Read data from MSR only if required
+	if p.c1CoreStateResidency || p.c6CoreStateResidency || p.processorBusyCycles || p.coreCPUTemperature ||
+		p.processorBusyFrequency {
+		err := p.msr.OpenAndReadMsr(cpuID)
+		if err != nil {
+			// In case of an error exit the function. All metrics past this point are dependant on MSR.
+			p.Log.Debugf("error while reading msr: %v", err)
+			return
+		}
+	}
+
+	if p.coreCPUTemperature {
+		p.addCoreCPUTemperatureMetric(cpuID, acc)
+	}
+	if !p.skipFirstIteration {
+		if p.c6CoreStateResidency {
+			p.addC6StateResidencyMetric(cpuID, acc)
+		}
+
+		if p.processorBusyCycles {
+			p.addProcessorBusyCyclesMetric(cpuID, acc)
+		}
+
+		if p.c1CoreStateResidency {
+			p.addC1StateResidencyMetric(cpuID, acc)
+		}
+	}
+	// processorBusyFrequency metric does some calculations inside that are required in another plugin cycle.
+	if p.processorBusyFrequency {
+		p.addProcessorBusyFrequencyMetric(cpuID, acc)
+	}
+}
+
+func (p *PowerStat) addCPUFrequencyMetric(cpuID string, acc telegraf.Accumulator) {
+	frequency, _, err := p.msr.RetrieveCPUFrequencyForCore(cpuID)
+
+	// In case of an error leave func
+	if err != nil {
+		p.Log.Debugf("error while reading file: %v", err)
+		return
+	}
+
+	cpu := p.cpuInfo[cpuID]
+	tags := map[string]string{
+		"package_id": cpu.PhysicalID,
+		"core_id":    cpu.CoreID,
+		"cpu_id":     cpu.CPUID,
+	}
+
+	fields := map[string]interface{}{
+		"name":  "cpu_frequency",
+		"unit":  "MHz",
+		"value": services.RoundFloatToNearestTwoDecimalPlaces(services.ConvertKiloHertzToMegaHertz(frequency)),
```

Comment:
I think this is not so nice as querying the data will be hard. Looking up `cpu_frequencies` might be costly as you can only constrain by field value which means you have to walk all rows in the table. How about either making the "name" and "unit" a tag **or**
```suggestion
		"cpu_frequency_mhz": services.RoundFloatToNearestTwoDecimalPlaces(services.ConvertKiloHertzToMegaHertz(frequency)),
```
and maybe unit a tag?

Reply:
Same for all other instances.

Reply:
Excellent point. Fixed.